lib/verification-reminders.js

```diff
@@ -0,0 +1,137 @@
+  }, { keys: [], intervals: [] });
+
+  /**
+   * @typedef {Object} VerificationReminders
+   * @property {Array} keys
+   * @property {Function} create
+   * @property {Function} delete
+   * @property {Function} process
+   *
+   * Each method below returns a promise that resolves to an object,
+   * the shape of which is determined by config. If config has settings
+   * for `firstInterval` and `secondInterval` (as at time of writing),
+   * the shape of those objects would be `{ first, second }`.
+   */
+  return {
+    keys: keys.slice(),
+
+    /**
+     * Create verification reminder records for an account.
+     *
+     * @param {String} uid
+     * @returns {Promise} - Each property on the resolved object will be the number
+     *                      of elements added to that sorted set, i.e. the result of
+     *                      [`redis.zadd`](https://redis.io/commands/zadd).
+     */
+    async create (uid) {
+      try {
+        if (rolloutRate <= 1 && Math.random() < rolloutRate) {
+          const now = Date.now();
+          const result = await P.props(keys.reduce((result, key) => {
+            result[key] = redis.zadd(key, now, uid);
+            return result;
+          }, {}));
+          log.info('verificationReminders.create', { uid });
+          return result;
+        }
+      } catch (err) {
+        log.error('verificationReminders.create.error', { err, uid });
+        throw err;
+      }
+    },
+
+    /**
+     * Delete verification reminder records for an account.
+     *
+     * @param {String} uid
+     * @returns {Promise} - Each property on the resolved object will be the number of
+     *                      elements removed from that sorted set, i.e. the result of
+     *                      [`redis.zrem`](https://redis.io/commands/zrem).
+     */
+    async delete (uid) {
+      try {
+        const result = await P.props(keys.reduce((result, key) => {
+          result[key] = redis.zrem(key, uid);
+          return result;
+        }, {}));
+        log.info('verificationReminders.delete', { uid });
+        return result;
+      } catch (err) {
+        log.error('verificationReminders.delete.error', { err, uid });
+        throw err;
+      }
+    },
+
+    /**
+     * Read and remove all verification reminders that have
+     * ticked past the expiry intervals set in config.
+     *
+     * @returns {Promise} - Each property on the resolved object will be an array of uids that
+     *                      were found to have ticked past the relevant expiry interval, i.e.
+     *                      the result of [`redis.zrangebyscore`](https://redis.io/commands/zrangebyscore).
+     */
+    async process () {
+      try {
+        const now = Date.now();
+        return await P.props(keys.reduce((result, key, index) => {
+          const cutoff = now - intervals[index];
+          result[key] = redis.zrangebyscore(key, 0, cutoff);
+          setImmediate(async () => {
```

Comment:
What is the setImmediate used here for? I'm trying to understand the flow control and am confused.

Reply:
`setImmediate` is a hack to make the inner function `async`, so that the outer `reduce` doesn't have to wait on the state of a promise and therefore run the operations serially rather than concurrently. If it's too difficult to follow though, I can get rid, I was just trying to find a way not to have to write the promise-handling code manually.

Reply:
>  I can get rid, I was just trying to find a way not to have to write the promise-handling code manually.

Nah, keep it. Can you leave a comment about why it's there so other readers understand why it's there too?